### DISCUSSION TITLE
KEYWORDS with or without quotes in ANTLR4

### DISCUSSION THREAD

**Discussion by techpranav** [antlr/antlr4#4173] [Q&A]
Hi,

I have a requirement like grammar should parse all the keywords irrespective of quotes.
e.g.:
 ADD_ :  'add';
 DELETE_: ''delete;
 INSERT_: 'insert';

KEYWORDS: ( ADD_ | DELETE_ | INSERT_ ); 
TYPE_ : 'type';

stmt : (KEYWORDS) TYPE_ ANY_STRING;

ANY_STRING :  (ALPHA | DIGIT | SYMBOL)+;
SYMBOL : [\p{P}\p{S}]; 
ALPHA : [\p{L}\p{Nl}]; 
DIGIT : [\p{Nd}]; 

===============================
Below are some sample valid statements -
add type Test;
"add" type Test;
'add' ' type' Test;
add "type" Test;

===============================

How can I implement this logic ?
Note: I have bunch of queries like this , so looking for some generic solution?

**Comment by jimidle** [comment]
I am not sure if this is your entire grammar or not, but your lexer does not handle quotes at all as it stands. Here are some generic observations:

1. Do not make your rule for alternative keywords a lexer rule, it should be a grammar rule.
2. You don't seem to need actual quoted strings, so this makes it fairly easy. If you need quoted strings as well, then you are going to have different issues and probably want to use lexer modes.
3. Your grammar as shown does not handle quotes or characters that are not otherwise matched in lexer rules.
4. Don't use _ as a suffix - it makes things less readable.

## Rewriting the above:
 
```
stmt       : keywords TYPE any_string ;
keywords   : words | QUOTE words QUOTE ;
words      : ADD | DELETE | INSERT ;
any_string : SYMBOL | ALPHA | DIGIT ;

QUOTE  : '\'' | '"' ;
TYPE   : 'type';
ADD    : 'add';
DELETE : ''delete;
INSERT : 'insert';

SYMBOL : [\p{P}\p{S}];
ALPHA  : [\p{L}\p{Nl}];
DIGIT  : [\p{Nd}];

BAD : . -> skip() ;
```

**Reply by techpranav** [reply]
Hi,

It's working in simple standalone Test.g4 but when I implement a similar approach for my original code it does not match.

![image](https://user-images.githubusercontent.com/30039186/224738754-bfe78c35-b278-49b3-84f9-b21dfdbb8532.png)

**Reply by techpranav** [reply]
Hi,

`grammar Test;

options {
    caseInsensitive = true;
}
stmt       : any_string TYPE keywords ;
any_string : arg_string | QUOTE arg_string QUOTE;

arg_string : (SYMBOL | ALPHA | DIGIT)+ ;


keywords   : words | QUOTE words QUOTE ;
words      : ADD | DELETE | INSERT ;
QUOTE  : '\'' | '"' ;
TYPE   : 'type';
ADD    : 'add';
DELETE : 'delete';
INSERT : 'insert';

SYMBOL : [\p{P}\p{S}]+;
ALPHA  : [\p{L}\p{Nl}]+;
DIGIT  : [\p{Nd}]+;

BAD : . -> skip;
`

Output:

![image](https://user-images.githubusercontent.com/30039186/224741030-a613f049-8168-43aa-9645-c0546b922e8c.png)


Why this is failing for same grammar?
